### PR TITLE
[WIP] Gradient collector class: an iterator returning model weights

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

Work in progress.

This PR aims to propose a single class to abstract data collection and gradient computation.

This solution might not be very useful in a local training setting but could be use as a base worker for distributed approach like RAPID.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/gradient_collector.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        device: torch.device = "cpu",
+    ):
+
+        self.device = device
+        self.updates_per_batch = updates_per_batch
+
+        # Get Actor Critic instance
+        # Could be instantiated passing class + params if necessary
+        self.policy = policy  # TODO: do I really need the policy if we can get the params from objective ?
+        self.policy_params = TensorDict(dict(self.policy.named_parameters()), [])
```
Comment (vmoens) [inline torchrl/gradient_collector.py:55]: we should add the buffers to this

File: torchrl/gradient_collector.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        self.critic = critic  # TODO: do I really need the critic if we can get the params from objective ?
+        self.critic_params = TensorDict(dict(self.critic.named_parameters()), [])
```
Comment (vmoens) [inline torchrl/gradient_collector.py:57]: ditto

File: torchrl/gradient_collector.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+        # Get loss instance
+        # Could be instantiated passing class + params if necessary
+        self.objective = objective
+        self.objective_params = TensorDict(dict(self.objective.named_parameters()), [])
```
Comment (vmoens) [inline torchrl/gradient_collector.py:62]: I expect those params to be the same as policy and critic

File: torchrl/gradient_collector.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            return None
+
+    def update_policy_weights_(
+            self,
+            policy_weights: Optional[TensorDictBase] = None,
+            critic_weights: Optional[TensorDictBase] = None,
+            objective_weights: Optional[TensorDictBase] = None,
+    ) -> None:
+
+        # TODO: is this the right way to do it ?
```
Comment (vmoens) [inline torchrl/gradient_collector.py:108]: yeah I think it should work

File: torchrl/gradient_collector.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        # Update policy and critic
+        self.objective_params.apply(lambda x: x.data).update_(objective_weights)
+        self.policy_params.apply(lambda x: x.data).update_(policy_weights)
+        self.critic_params.apply(lambda x: x.data).update_(critic_weights)
+
+        for name, param in self.objective.named_parameters():
```
Comment (vmoens) [inline torchrl/gradient_collector.py:114]: in theory your objective params and buffers should match in identity those of the policy and critic
One way to make sure that it's the case would be to register only the objective params and buffers, and during this (in `__init__`) check that they match in identity those of the policy and critic, smth like
```python
if set(policy.parameters()).union(set(critic.parameters())) != objective.parameters():
    raise ValueError
```

File: torchrl/gradient_collector.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                self.optimizer.zero_grad()
+                loss_sum.backward()
+
+                # Get gradients as a Tensordict
+                objective_params = TensorDict(dict(self.objective.named_parameters()), [])
+                objective_grads = objective_params.apply(lambda x: x.grad)
+                actor_params = TensorDict(dict(self.policy.named_parameters()), [])
+                actor_grads = actor_params.apply(lambda x: x.grad)
+                critic_params = TensorDict(dict(self.critic.named_parameters()), [])
+                critic_grads = critic_params.apply(lambda x: x.grad)
```
Comment (vmoens) [inline torchrl/gradient_collector.py:157]:
```suggestion
                objective_params = TensorDict(dict(self.objective.named_parameters()), [])
                objective_grads = objective_params.apply(lambda x: x.grad)
```

Only the first two lines should be necessary

Not sure how that would work if the objective_params also contain buffers. Maybe we should do

```
objective_grads = objective_params.apply(lambda x: x.grad if x.grad is not None else torch.zeros_like(x))
```

but it could be unsafe (e.g. if by mistake a param is ignored)

**Review by vmoens — COMMENTED** [review]
Some early thoughts
